src/mongoc/mongoc-cluster.c

```diff
@@ -1904,559 +962,436 @@ _mongoc_cluster_auth_node (mongoc_cluster_t      *cluster,
-      }
+   mongoc_cluster_disconnect_node(cluster, server_id);
 
-      if (!stream) {
-         stream = _mongoc_client_create_stream (cluster->client, &host, error);
+   return NULL;
+}
 
-         if (!stream) {
-            MONGOC_WARNING("Failed connection to %s", host.host_and_port);
-            continue;
-         }
-      }
 
-      _mongoc_cluster_node_init(&cluster->nodes[i]);
+static void
+_mongoc_cluster_stream_dtor (void *stream_,
+                             void *ctx_)
+{
+   mongoc_stream_destroy ((mongoc_stream_t *)stream_);
+}
 
-      cluster->nodes[i].host = host;
-      cluster->nodes[i].index = i;
-      cluster->nodes[i].stream = stream;
-      cluster->nodes[i].needs_auth = cluster->requires_auth;
+/*
+ *--------------------------------------------------------------------------
+ *
+ * mongoc_cluster_init --
+ *
+ *       Initializes @cluster using the @uri and @client provided. The
+ *       @uri is used to determine the "mode" of the cluster. Based on the
+ *       uri we can determine if we are connected to a single host, a
+ *       replicaSet, or a shardedCluster.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       @cluster is initialized.
+ *
+ *--------------------------------------------------------------------------
+ */
 
-      if (!_mongoc_cluster_ismaster(cluster, &cluster->nodes[i], error)) {
-         _mongoc_cluster_node_destroy(&cluster->nodes[i]);
-         continue;
-      }
+void
+mongoc_cluster_init (mongoc_cluster_t   *cluster,
+                     const mongoc_uri_t *uri,
+                     void               *client)
+{
+   ENTRY;
 
-      if (!cluster->nodes[i].replSet ||
-          !!strcmp (cluster->nodes[i].replSet, replSet)) {
-         MONGOC_INFO ("%s: Got replicaSet \"%s\" expected \"%s\".",
-                      host.host_and_port,
-                      cluster->nodes[i].replSet ? cluster->nodes[i].replSet : "(null)",
-                      replSet);
-         _mongoc_cluster_node_destroy (&cluster->nodes[i]);
-         continue;
-      }
+   bson_return_if_fail(cluster);
+   bson_return_if_fail(uri);
 
-      if (cluster->nodes[i].needs_auth) {
-         if (!_mongoc_cluster_auth_node (cluster, &cluster->nodes[i], error)) {
-            _mongoc_cluster_node_destroy (&cluster->nodes[i]);
-            goto CLEANUP;
-         }
-         cluster->nodes[i].needs_auth = false;
-      }
+   memset (cluster, 0, sizeof *cluster);
 
-      if (-1 == (ping = _mongoc_cluster_ping_node (cluster,
-                                                   &cluster->nodes[i],
-                                                   error))) {
-         MONGOC_INFO("%s: Lost connection during ping.",
-                     host.host_and_port);
-         _mongoc_cluster_node_destroy (&cluster->nodes[i]);
-         continue;
-      }
+   cluster->uri = mongoc_uri_copy(uri);
+   cluster->client = client;
+   cluster->max_msg_size = 1024 * 1024 * 48;
+   cluster->max_bson_size = 1024 * 1024 * 16;
```

Comment:
max_bson_size should be set to 4194304, and max_message_size to max_bson_size * 2